common/prettify.pm

```diff
@@ -353,6 +353,194 @@ sub Normal ($)
+package Prettify::Failed_Tests_HTML;
+
+use strict;
+use warnings;
+
+use FileHandle;
+
+###############################################################################
+
+sub new ($)
+{
+    my $proto = shift;
+    my $class = ref ($proto) || $proto;
+    my $self = {};
+    my $basename = shift;
+    my $buildname = shift;
+    my $rev_link = shift;
+    my $log_prefix = shift;
+
+    my $filename = $log_prefix . "_Failed_Tests_By_Build.html";
+
+    $basename =~ s/^.*\///;
+
+    $self->{FULLHTML} = "$buildname/$basename" . "_Full.html";
+    $self->{ERROR_COUNTER} = 0;
+    $self->{WARNING_COUNTER} = 0;
+    $self->{SECTION_COUNTER} = 0;
+    $self->{SUBSECTION_COUNTER} = 0;
+    $self->{TITLE} = "Failed Test Brief Log By Build";
+    $self->{GIT_CHECKEDOUT_OPENDDS} = "unknown";
+    $self->{REV_LINK} = $rev_link;
+
+    unless (-e $filename) {
+        my $file_handle = new FileHandle ($filename, 'w');
+        print {$file_handle} "<h1>$self->{TITLE}</h1>\n";
+    }
+
+    $self->{FH} = new FileHandle ($filename, '>>');
+    $self->{FILENAME} = $filename;
+    $self->{BUILDNAME} = $buildname;
+
+    bless ($self, $class);
+    return $self;
+}
+
+sub Header ()
+{
+    my $self = shift;
+
+    if (defined $self->{LAST_SECTION} && $self->{LAST_SECTION} eq 'Test') {
+        print {$self->{FH}} "<html>\n";
+        print {$self->{FH}} "<body bgcolor=\"white\">\n";
+    }
+}
+
+sub Footer ()
+{
+    my $self = shift;
+
+    if (defined $self->{LAST_SECTION} && $self->{LAST_SECTION} eq 'Test') {
+        # In the case where there was no errors or warnings, output a note
+        if ($self->{ERROR_COUNTER} == 0 && $self->{WARNING_COUNTER} == 0) {
+            print {$self->{FH}} "No Errors or Warnings detected<br>\n";
+        }
+
+        print {$self->{FH}} "</body>\n";
+        print {$self->{FH}} "</html>\n";
+    }
+}
+
+sub Section ($)
+{
+    my $self = shift;
+    my $s = shift;
+
+    # Escape any '<' or '>' signs
+    $s =~ s/</&lt;/g;
+    $s =~ s/>/&gt;/g;
+
+    my $counter = ++$self->{SECTION_COUNTER};
+
+    # Save for later use
+
+    $self->{LAST_SECTION} = $s;
+}
+
+sub Description ($)
+{
+    my $self = shift;
+
+    # Ignore
+}
+
+sub Timestamp ($)
+{
+    my $self = shift;
+    # Ignore
+}
+
+sub Subsection ($)
+{
+    my $self = shift;
+    my $s = shift;
+
+    # Escape any '<' or '>' signs
+    $s =~ s/</&lt;/g;
+    $s =~ s/>/&gt;/g;
+
+    my $counter = ++$self->{SUBSECTION_COUNTER};
+
+    # Save for later use
+
+    $self->{LAST_SUBSECTION} = $s;
+}
+
+sub Print_Sections ()
+{
+    my $self = shift;
+
+    if (defined $self->{LAST_SECTION} && defined $self->{LAST_SUBSECTION} && $self->{LAST_SECTION} eq 'Test') {
+        if (defined $self->{BUILDNAME}) {
+            print {$self->{FH}} "<hr><h2>$self->{BUILDNAME}</h2>\n";
+            my $rev = substr($self->{GIT_CHECKEDOUT_OPENDDS}, 0, 8);
+            if ($rev ne "unknown") {
+                print {$self->{FH}} "Rev: <a href=$self->{REV_LINK}/$rev>$rev</a><hr>\n";
```

Comment:
Two things: it shouldn't be a link if `diffroot` wasn't set and it the `diffroot` might have already slash at the end like it does on our score board.

```suggestion
                my $rev_line = "Rev: ";
                if (length($self->{REV_LINK})) {
                    $rev_line .= "<a href=$self->{REV_LINK}";
                    $rev_line =~ s/\/$//g;
                    $rev_line .= "/$rev>";
                }
                $rev_line .= $rev;
                if (length($self->{REV_LINK})) {
                    $rev_line .= "</a>";
                }
                print {$self->{FH}} "$rev_line<hr>\n";
```

Other than that everything looks alright.

Reply:
Thank you, Fred